File: mobile/src/main/java/com/windscribe/mobile/welcome/WelcomeActivity.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.mobile.welcome;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.transition.Slide;
import android.view.Gravity;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.app.ShareCompat;
import androidx.core.content.FileProvider;
import androidx.core.view.GravityCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import com.windscribe.mobile.R;
import com.windscribe.mobile.alert.UnknownErrorAlert;
import com.windscribe.mobile.alert.UnknownErrorAlert.LoginAttemptFailedAlertInterface;
import com.windscribe.mobile.base.BaseActivity;
import com.windscribe.mobile.custom_view.ErrorFragment;
import com.windscribe.mobile.custom_view.ProgressFragment;
import com.windscribe.mobile.di.ActivityModule;
import com.windscribe.mobile.di.DaggerActivityComponent;
import com.windscribe.mobile.welcome.fragment.FragmentCallback;
import com.windscribe.mobile.welcome.fragment.LoginFragment;
import com.windscribe.mobile.welcome.fragment.NoEmailAttentionFragment;
import com.windscribe.mobile.welcome.fragment.SignUpFragment;
import com.windscribe.mobile.welcome.fragment.WelcomeActivityCallback;
import com.windscribe.mobile.welcome.fragment.WelcomeFragment;
import com.windscribe.mobile.windscribe.WindscribeActivity;
import com.windscribe.vpn.Windscribe;
import com.windscribe.vpn.constants.NetworkKeyConstants;

import java.io.File;

import javax.inject.Inject;

import butterknife.ButterKnife;


public class WelcomeActivity extends BaseActivity
        implements FragmentCallback, WelcomeView, LoginAttemptFailedAlertInterface {

    @Inject
    WelcomePresenter presenter;

    private final int REQUEST_PERMISSION_CODE = 201;

    private SoftInputAssist softInputAssist;

    public static Intent getStartIntent(Context context) {
        return new Intent(context, WelcomeActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_welcome);
        DaggerActivityComponent.builder().activityModule(new ActivityModule(this, this))
                .applicationComponent(Windscribe.getAppContext()
                        .getApplicationComponent()).build().inject(this);
        ButterKnife.bind(this);
        addStartFragment();

        Uri uri = ActivityCompat.getReferrer(this);
        presenter.printStartUri(uri);

    }

    @Override
    protected void onResume() {
        super.onResume();
        softInputAssist.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        softInputAssist.onPause();
    }

    @Override
    protected void onDestroy() {
        softInputAssist.onDestroy();
        presenter.onDestroy();
        super.onDestroy();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
            @NonNull int[] grantResults) {
        if (requestCode == REQUEST_PERMISSION_CODE) {
            if (permissionGranted()) {
                presenter.exportLog();
            } else {
                showToast("Please provide storage permission");
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    public void clearInputErrors() {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
        if (fragment instanceof WelcomeActivityCallback) {
            ((WelcomeActivityCallback) fragment).clearInputErrors();
        }
    }

    @Override
    public void contactSupport() {
        openURLInBrowser(NetworkKeyConstants.getWebsiteLink(NetworkKeyConstants.URL_HELP_ME));
    }

    @Override
    public void exportLog() {
        if (permissionGranted()) {
            presenter.exportLog();
        } else {
            askForPermission();
        }
    }

    @Override
    public void goToSignUp() {
        SignUpFragment signUpFragment = new SignUpFragment(false);
        int direction = GravityCompat
                .getAbsoluteGravity(GravityCompat.END, getResources().getConfiguration().getLayoutDirection());
        signUpFragment.setEnterTransition(new Slide(direction).addTarget(R.id.sign_up_container));
        replaceFragment(signUpFragment, true);
    }

    @Override
    public void gotoHomeActivity(boolean clearTop) {
        Intent startIntent = new Intent(this, WindscribeActivity.class);
        if (clearTop) {
            startIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        }
        startActivity(startIntent);
        finish();
    }

    @Override
    public void hideSoftKeyboard() {
        InputMethodManager imm = (InputMethodManager)
                getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.hideSoftInputFromWindow(getWindow()
                    .getDecorView().getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }

    @Override
    public void launchShareIntent(File file) {
        Uri fileUri = FileProvider.getUriForFile(
                this,
                "com.windscribe.vpn.provider",
                file);
        ShareCompat.IntentBuilder.from(this)
                .setType("*/*")
                .setStream(fileUri).startChooser();
    }

    @Override
    public void onAccountClaimButtonClick(String username, String password, String email, boolean ignoreEmptyEmail) {
        presenter.startAccountClaim(username, password, email, ignoreEmptyEmail);
    }

    @Override
    public void onBackButtonPressed() {
        onBackPressed();
    }

    @Override
    public void onBackPressed() {
        presenter.onBackPressed();
        super.onBackPressed();
    }

    @Override
    public void onContinueWithOutAccountClick() {
        presenter.startGhostAccountSetup();
    }

    @Override
    public void onForgotPasswordClick() {
        openURLInBrowser(NetworkKeyConstants.getWebsiteLink(NetworkKeyConstants.URL_FORGOT_PASSWORD));
    }

    @Override
    public void onLoginButtonClick(String username, String password, String twoFa) {
        presenter.startLoginProcess(username, password, twoFa);
    }

    @Override
    public void onLoginClick() {
        LoginFragment loginFragment = new LoginFragment();
        int direction = GravityCompat
                .getAbsoluteGravity(GravityCompat.END, getResources().getConfiguration().getLayoutDirection());
        loginFragment.setEnterTransition(new Slide(direction).addTarget(R.id.login_container));
        replaceFragment(loginFragment, true);
    }

    @Override
    public void onSignUpButtonClick(String username, String password, String email, boolean ignoreEmptyEmail) {
        if (ignoreEmptyEmail) {
            getSupportFragmentManager().popBackStack();
        }
        presenter.startSignUpProcess(username, password, email, ignoreEmptyEmail);
    }

    @Override
    public void onSkipToHomeClick() {
        Intent startIntent = new Intent(this, WindscribeActivity.class);
        startActivity(startIntent);
        finish();
    }

    @Override
    public void prepareUiForApiCallFinished() {
        Fragment progressFragment = getSupportFragmentManager().findFragmentById(R.id.progress_container);
        if (progressFragment instanceof ProgressFragment | progressFragment instanceof NoEmailAttentionFragment) {
            getSupportFragmentManager().popBackStack();
        }
        Fragment mainFragment = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
        if (mainFragment instanceof NoEmailAttentionFragment) {
            getSupportFragmentManager().popBackStack();
        }
    }

    @Override
    public void prepareUiForApiCallStart() {
        ProgressFragment.getInstance().add(this, R.id.progress_container, true);
    }

    public void replaceFragment(Fragment fragment, boolean addToBackStack) {
        FragmentTransaction transaction = getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.fragment_container, fragment);
        if (addToBackStack) {
            transaction.addToBackStack(fragment.getClass().getName());
        }
        transaction.commit();
    }

    @Override
    public void setEmailError(String errorMessage) {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
        if (fragment instanceof SignUpFragment) {
            ((SignUpFragment) fragment).setEmailError(errorMessage);
        }
    }

    @Override
    public void setFaFieldsVisibility(int visibility) {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
        if (fragment instanceof LoginFragment) {
            ((LoginFragment) fragment).setTwoFaVisibility(visibility);
        }
    }

    @Override
    public void setLoginRegistrationError(String error) {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
        if (fragment instanceof WelcomeActivityCallback) {
            ((WelcomeActivityCallback) fragment).setLoginError(error);
        }
    }

    @Override
    public void setPasswordError(String error) {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
        if (fragment instanceof WelcomeActivityCallback) {
            ((WelcomeActivityCallback) fragment).setPasswordError(error);
        }
    }

    @Override
    public void setTwoFaError(String errorMessage) {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
        if (fragment instanceof LoginFragment) {
            ((LoginFragment) fragment).setTwoFaError(errorMessage);
        }
    }

    @Override
    public void setUsernameError(String error) {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
        if (fragment instanceof WelcomeActivityCallback) {
            ((WelcomeActivityCallback) fragment).setUsernameError(error);
        }
    }

    public void setWindow() {
        int statusBarColor = getResources().getColor(android.R.color.transparent);
        getWindow().getDecorView()
                .setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
        getWindow().setStatusBarColor(statusBarColor);
    }

    @Override
    public void showError(String error) {
        ErrorFragment.getInstance().add(error, this, R.id.fragment_container, true);
    }

    @Override
    public void showFailedAlert(String error) {
        runOnUiThread(() -> {
            UnknownErrorAlert unknownErrorAlert = UnknownErrorAlert.newInstance(error);
            unknownErrorAlert.show(getSupportFragmentManager(), "failed_login");
        });
    }

    @Override
    public void showNoEmailAttentionFragment(String username, String password, boolean accountClaim,
            boolean isUserPro) {
        NoEmailAttentionFragment noEmailAttentionFragment = new NoEmailAttentionFragment(accountClaim, username,
                password, isUserPro);
        noEmailAttentionFragment
                .setEnterTransition(new Slide(Gravity.BOTTOM).addTarget(R.id.email_fragment_container));
        getSupportFragmentManager().beginTransaction()
                .add(R.id.fragment_container, noEmailAttentionFragment)
                .addToBackStack(noEmailAttentionFragment.getClass().getName())
                .commit();
    }

    public void showToast(String error) {
        Toast.makeText(this, error, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void updateCurrentProcess(String call) {
        runOnUiThread(() -> {
            Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.progress_container);
            if (fragment instanceof ProgressFragment) {
                ((ProgressFragment) fragment).updateProgressStatus(call);
            }
        });

    }

    void addStartFragment() {
        String startFragmentName = getIntent().getStringExtra("startFragmentName");
        boolean skipToHome = getIntent().getBooleanExtra("skipToHome", false);
        Fragment fragment;
        if (startFragmentName != null && startFragmentName.equals("Login")) {
            fragment = new LoginFragment();
            softInputAssist = new SoftInputAssist(this, new int[]{R.id.forgot_password, R.id.page_description});
        } else if (startFragmentName != null && startFragmentName.equals("SignUp")) {
            softInputAssist = new SoftInputAssist(this, new int[]{R.id.forgot_password, R.id.page_description});
            fragment = new SignUpFragment(false);
        } else if (startFragmentName != null && startFragmentName.equals("AccountSetUp")) {
            softInputAssist = new SoftInputAssist(this,
                    new int[]{R.id.forgot_password, R.id.page_description, R.id.set_up_later_button});
            boolean proAccount = presenter.isUserPro();
            fragment = new SignUpFragment(proAccount);
        } else {
            softInputAssist = new SoftInputAssist(this, new int[]{R.id.forgot_password, R.id.page_description});
            fragment = new WelcomeFragment();
        }
        Bundle bundle = new Bundle();
        bundle.putString("startFragmentName", startFragmentName);
        bundle.putBoolean("skipToHome", skipToHome);
        fragment.setArguments(bundle);
        int direction = GravityCompat
                .getAbsoluteGravity(GravityCompat.END, getResources().getConfiguration().getLayoutDirection());
        fragment.setEnterTransition(new Slide(direction).addTarget(R.id.welcome_container));
        replaceFragment(fragment, false);
    }

    void askForPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_PERMISSION_CODE);
        }
    }

    boolean permissionGranted() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            return checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                    == PackageManager.PERMISSION_GRANTED;
        } else {
            return true;
        }
    }
}

File: mobile/src/main/java/com/windscribe/mobile/welcome/WelcomeActivity.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.mobile.welcome

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.transition.Slide
import android.view.Gravity
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.core.app.ShareCompat
import androidx.core.content.FileProvider
import androidx.core.view.GravityCompat
import androidx.fragment.app.Fragment
import butterknife.ButterKnife
import com.windscribe.mobile.R
import com.windscribe.mobile.base.BaseActivity
import com.windscribe.mobile.di.ActivityModule
import com.windscribe.mobile.di.DaggerActivityComponent
import com.windscribe.mobile.dialogs.ErrorDialog
import com.windscribe.mobile.dialogs.ProgressDialog
import com.windscribe.mobile.dialogs.UnknownErrorDialog
import com.windscribe.mobile.dialogs.UnknownErrorDialogCallback
import com.windscribe.mobile.welcome.fragment.EmergencyConnectFragment
import com.windscribe.mobile.welcome.fragment.FragmentCallback
import com.windscribe.mobile.welcome.fragment.LoginFragment
import com.windscribe.mobile.welcome.fragment.NoEmailAttentionFragment
import com.windscribe.mobile.welcome.fragment.SignUpFragment
import com.windscribe.mobile.welcome.fragment.WelcomeActivityCallback
import com.windscribe.mobile.welcome.fragment.WelcomeFragment
import com.windscribe.mobile.welcome.state.EmergencyConnectUIState
import com.windscribe.mobile.welcome.viewmodal.EmergencyConnectViewModal
import com.windscribe.mobile.windscribe.WindscribeActivity
import com.windscribe.vpn.Windscribe.Companion.appContext
import com.windscribe.vpn.constants.NetworkKeyConstants
import com.windscribe.vpn.constants.NetworkKeyConstants.getWebsiteLink
import java.io.File
import javax.inject.Inject

class WelcomeActivity : BaseActivity(), FragmentCallback, WelcomeView, UnknownErrorDialogCallback {

    @Inject
    lateinit var presenter: WelcomePresenter

    @Inject
    lateinit var emergencyConnectViewModal: Lazy<EmergencyConnectViewModal>

    private val requestLocationPermissionCode = 201
    private var softInputAssist: SoftInputAssist? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_welcome)
        DaggerActivityComponent.builder().activityModule(ActivityModule(this, this))
            .applicationComponent(
                appContext.applicationComponent
            ).build().inject(this)
        ButterKnife.bind(this)
        addStartFragment()
    }

    override fun onResume() {
        super.onResume()
        softInputAssist?.onResume()
    }

    override fun onPause() {
        super.onPause()
        softInputAssist?.onPause()
    }

    override fun onDestroy() {
        softInputAssist?.onDestroy()
        presenter.onDestroy()
        super.onDestroy()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults: IntArray
    ) {
        if (requestCode == requestLocationPermissionCode) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                presenter.exportLog()
            } else {
                showToast("Please provide storage permission")
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun clearInputErrors() {
        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (fragment is WelcomeActivityCallback) {
            (fragment as WelcomeActivityCallback).clearInputErrors()
        }
    }

    override fun contactSupport() {
        val intent = Intent(Intent.ACTION_SEND)
        intent.setType("plain/text")
        intent.putExtra(Intent.EXTRA_EMAIL, arrayOf("helpdesk@windscribe.com"))
        intent.putExtra(Intent.EXTRA_SUBJECT, "Restrictive Network Detected")
        intent.putExtra(Intent.EXTRA_TEXT, "")
        presenter.getLogUri()?.let {
            val fileUri = FileProvider.getUriForFile(this, "com.windscribe.vpn.provider", it)
            intent.putExtra(Intent.EXTRA_STREAM, fileUri)
        }
       if(intent.resolveActivity(packageManager) != null){
           startActivity(Intent.createChooser(intent, "Select Email Provider"))
       }
    }

    override fun exportLog() {
        presenter.exportLog()
    }

    override fun goToSignUp() {
        val signUpFragment = SignUpFragment.newInstance(false)
        val direction = GravityCompat.getAbsoluteGravity(
            GravityCompat.END, resources.configuration.layoutDirection
        )
        signUpFragment.enterTransition = Slide(direction).addTarget(R.id.sign_up_container)
        replaceFragment(signUpFragment, true)
    }

    override fun gotoHomeActivity(clearTop: Boolean) {
        if (emergencyConnectViewModal.value.uiState.value != EmergencyConnectUIState.Disconnected) {
            emergencyConnectViewModal.value.disconnect()
        }
        val startIntent = Intent(this, WindscribeActivity::class.java)
        if (clearTop) {
            startIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        }
        startActivity(startIntent)
        finish()
    }

    override fun hideSoftKeyboard() {
        val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(
            window.decorView.windowToken, InputMethodManager.HIDE_NOT_ALWAYS
        )
    }

    override fun launchShareIntent(file: File) {
        val fileUri = FileProvider.getUriForFile(
            this, "com.windscribe.vpn.provider", file
        )
        ShareCompat.IntentBuilder.from(this).setType("*/*").setStream(fileUri).startChooser()
    }

    override fun onAccountClaimButtonClick(
        username: String, password: String, email: String, ignoreEmptyEmail: Boolean, voucherCode: String
    ) {
        presenter.startAccountClaim(username, password, email, ignoreEmptyEmail, voucherCode)
    }

    override fun onBackButtonPressed() {
        onBackPressed()
    }

    override fun onBackPressed() {
        presenter.onBackPressed()
        super.onBackPressed()
    }

    override fun onContinueWithOutAccountClick() {
        presenter.startGhostAccountSetup()
    }

    override fun onForgotPasswordClick() {
        openURLInBrowser(getWebsiteLink(NetworkKeyConstants.URL_FORGOT_PASSWORD))
    }

    override fun onLoginButtonClick(username: String, password: String, twoFa: String) {
        presenter.startLoginProcess(username, password, twoFa)
    }

    override fun onEmergencyClick() {
        EmergencyConnectFragment.show(supportFragmentManager, R.id.fragment_container)
    }

    override fun onLoginClick() {
        val loginFragment = LoginFragment()
        val direction = GravityCompat.getAbsoluteGravity(
            GravityCompat.END, resources.configuration.layoutDirection
        )
        loginFragment.enterTransition = Slide(direction).addTarget(R.id.login_container)
        replaceFragment(loginFragment, true)
    }

    override fun onSignUpButtonClick(
        username: String,
        password: String,
        email: String,
        referralUsername: String,
        ignoreEmptyEmail: Boolean,
        voucherCode: String
    ) {
        if (ignoreEmptyEmail) {
            supportFragmentManager.popBackStack()
        }
        presenter.startSignUpProcess(
            username, password, email, referralUsername, ignoreEmptyEmail, voucherCode
        )
    }

    override fun onSkipToHomeClick() {
        val startIntent = Intent(this, WindscribeActivity::class.java)
        startActivity(startIntent)
        finish()
    }

    override fun prepareUiForApiCallFinished() {
        ProgressDialog.hide(this)
        val progressFragment = supportFragmentManager.findFragmentById(R.id.progress_container)
        if (progressFragment is NoEmailAttentionFragment) {
            supportFragmentManager.popBackStack()
        }
        val mainFragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (mainFragment is NoEmailAttentionFragment) {
            supportFragmentManager.popBackStack()
        }
    }

    override fun prepareUiForApiCallStart() {
        ProgressDialog.show(this)
    }

    private fun replaceFragment(fragment: Fragment, addToBackStack: Boolean) {
        val transaction =
            supportFragmentManager.beginTransaction().replace(R.id.fragment_container, fragment)
        if (addToBackStack) {
            transaction.addToBackStack(fragment.javaClass.name)
        }
        transaction.commit()
    }

    override fun setEmailError(errorMessage: String) {
        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (fragment is SignUpFragment) {
            fragment.setEmailError(errorMessage)
        }
    }

    override fun setFaFieldsVisibility(visible: Int) {
        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (fragment is LoginFragment) {
            fragment.setTwoFaVisibility(visible)
        }
    }

    override fun setLoginRegistrationError(error: String) {
        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (fragment is WelcomeActivityCallback) {
            (fragment as WelcomeActivityCallback).setLoginError(error)
        }
    }

    override fun setPasswordError(error: String) {
        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (fragment is WelcomeActivityCallback) {
            (fragment as WelcomeActivityCallback).setPasswordError(error)
        }
    }

    override fun setTwoFaError(errorMessage: String) {
        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (fragment is LoginFragment) {
            fragment.setTwoFaError(errorMessage)
        }
    }

    override fun setUsernameError(error: String) {
        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (fragment is WelcomeActivityCallback) {
            (fragment as WelcomeActivityCallback).setUsernameError(error)
        }
    }

    override fun setWindow() {
        val statusBarColor = resources.getColor(android.R.color.transparent)
        window.decorView.systemUiVisibility =
            View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        window.statusBarColor = statusBarColor
    }

    override fun showError(error: String) {
        ErrorDialog.show(this, error)
    }

    override fun showFailedAlert(error: String) {
        UnknownErrorDialog.show(this, error)
    }

    override fun showNoEmailAttentionFragment(
        username: String, password: String, accountClaim: Boolean, pro: Boolean, voucherCode: String
    ) {
        val noEmailAttentionFragment =
            NoEmailAttentionFragment(accountClaim, username, password, pro, voucherCode)
        noEmailAttentionFragment.enterTransition =
            Slide(Gravity.BOTTOM).addTarget(R.id.email_fragment_container)
        supportFragmentManager.beginTransaction()
            .replace(R.id.fragment_container, noEmailAttentionFragment)
            .addToBackStack(noEmailAttentionFragment.javaClass.name).commit()
    }

    override fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    override fun updateCurrentProcess(mCurrentCall: String) {
        val fragment = supportFragmentManager.findFragmentByTag(ProgressDialog.tag)
        if (fragment is ProgressDialog) {
            fragment.updateProgressStatus(mCurrentCall)
        }
    }

    private fun addStartFragment() {
        val startFragmentName = intent.getStringExtra("startFragmentName")
        val skipToHome = intent.getBooleanExtra("skipToHome", false)
        val fragment: Fragment
        if (startFragmentName != null && startFragmentName == "Login") {
            fragment = LoginFragment()
            softInputAssist = SoftInputAssist(this, intArrayOf(R.id.forgot_password))
        } else if (startFragmentName != null && startFragmentName == "SignUp") {
            softInputAssist = SoftInputAssist(this, intArrayOf(R.id.forgot_password))
            fragment = SignUpFragment.newInstance(false)
        } else if (startFragmentName != null && startFragmentName == "AccountSetUp") {
            softInputAssist =
                SoftInputAssist(this, intArrayOf(R.id.forgot_password, R.id.set_up_later_button))
            val proAccount = presenter.isUserPro
            fragment = SignUpFragment.newInstance(proAccount)
        } else {
            softInputAssist = SoftInputAssist(this, intArrayOf(R.id.forgot_password))
            fragment = WelcomeFragment()
        }
        val bundle = Bundle()
        bundle.putString("startFragmentName", startFragmentName)
        bundle.putBoolean("skipToHome", skipToHome)
        fragment.arguments = bundle
        val direction = GravityCompat.getAbsoluteGravity(
            GravityCompat.END, resources.configuration.layoutDirection
        )
        fragment.enterTransition = Slide(direction).addTarget(R.id.welcome_container)
        replaceFragment(fragment, false)
    }

    companion object {
        @JvmStatic
        fun getStartIntent(context: Context?): Intent {
            return Intent(context, WelcomeActivity::class.java)
        }
    }
}